code/controllers/subsystem/parallax.dm

```diff
@@ -19,32 +29,76 @@ SUBSYSTEM_DEF(parallax)
-		if(!istype(A))
+	//Swap the 2 lists
+	if(!length(currentrun))
+		//Nothing to process here
+		if(!length(queued))
+			return
+		var/temp = currentrun
+		currentrun = queued
+		queued = temp
+		current_run_pointer = 1
+		//Check client count
+		throttle_ghosts = throttle_ghost_pop && length(GLOB.clients) >= throttle_ghost_pop
+		throttle_all = throttle_all_pop && length(GLOB.clients) >= throttle_all_pop
+	//Begin processing the processing queue
+	while(current_run_pointer <= length(currentrun))
+		//Use a pointer, less wasted processing than removing from the list
+		var/client/C = currentrun[current_run_pointer]
+		//Increment the current list pointer, so we process the next element
+		current_run_pointer ++
+		//No client (Disconnected)
+		if(!C)
 			continue
-		for (A; isloc(A.loc) && !isturf(A.loc); A = A.loc);
-
-		if(A != C.movingmob)
-			if(C.movingmob != null)
-				C.movingmob.client_mobs_in_contents -= C.mob
-				UNSETEMPTY(C.movingmob.client_mobs_in_contents)
-			LAZYINITLIST(A.client_mobs_in_contents)
-			A.client_mobs_in_contents += C.mob
-			C.movingmob = A
-		if (MC_TICK_CHECK)
+		C?.parallax_update_queued = FALSE
+		//Do the parallax update (Move it to the correct location)
+		if ((throttle_ghosts && isobserver(C.mob)) || (throttle_all))
+			C?.mob?.hud_used?.freeze_parallax()
+		else
+			C?.mob?.hud_used?.update_parallax()
+		//Tick check to prevent overrunning
+		if(MC_TICK_CHECK)
 			return
-	currentrun = null
+	//Processing is completed, clear the list
+	currentrun.len = 0
+
+/datum/controller/subsystem/parallax/proc/on_mob_login(datum/source, mob/new_login)
+	//Register the required signals
+	RegisterSignal(new_login, COMSIG_PARENT_MOVED_RELAY, .proc/on_mob_moved)
+	RegisterSignal(new_login, COMSIG_MOB_LOGOUT, .proc/on_mob_logout)
+
+/datum/controller/subsystem/parallax/proc/on_mob_logout(mob/source)
+	UnregisterSignal(source, COMSIG_PARENT_MOVED_RELAY)
+	UnregisterSignal(source, COMSIG_MOB_LOGOUT)
+
+/datum/controller/subsystem/parallax/proc/on_mob_moved(mob/moving_mob, atom/parent, force)
+	update_client_parallax(moving_mob.client)
+
+//We need a client var for optimisation purposes
+/client
+	var/parallax_update_queued = FALSE
+	var/last_parallax_update_tick
+
+/datum/controller/subsystem/parallax/proc/update_client_parallax(client/updater, force = FALSE)
+	//Already queued for update
+	if(!updater || updater?.parallax_update_queued)
+		return
+	//If we haven't updated yet, instantly update
+	if (updater?.last_parallax_update_tick < times_fired || force)
+		if ((throttle_ghosts && isobserver(updater.mob)) || (throttle_all))
+			updater?.mob?.hud_used?.freeze_parallax()
+		else
+			updater?.mob?.hud_used?.update_parallax()
+		//Don't allow an instant update on the next fire, to maintain parallax_free_fire_delay_ticks fire per tick max
+		updater?.last_parallax_update_tick = times_fired + parallax_free_fire_delay_ticks
+		return
+	//Mark it as being queued
+	updater?.parallax_update_queued = TRUE
```

Comment:
Unless you sleep on this, client shouldn't be nulled

Reply:
I don't think this is true due to the way byond works

Reply:
Unless you sleep somewhere or run async client won't get deleted here, so one check is enough

Reply:
updater is known not to be null at the start of the proc, this should be fixed